### PR TITLE
Added gift match fields for donor

### PR DESCRIPTION
**Pull request by tarhata**
image

### CODE REVIEW

**Comment by tarhata** [comment]
I decided to move the gift matching language and fields to the bottom because they are optional and all the optional notes and steps are on the bottom as well. Figured it made sense to put them all near each other. If it's better to put the gift match fields near the `Promise` amount, then I think we should make it a checkbox like `Asked`

**Comment by tarhata** [comment]
The test fails, but I decided to commit it just so that you can see how I thought to approach it. I don't think I'm understanding how the pages are rendered and what to look out for in the tests, so maybe you can explain it to me.. :frowning:

File: sjfnw/fund/models.py [hunk 1]
```diff
@@ -288,11 +288,16 @@ class Donor(models.Model):
   received_afternext = models.PositiveIntegerField(default=0,
       verbose_name='Received - year after next')
   gift_notified = models.BooleanField(default=False)
+  match_expected = models.PositiveIntegerField(max_length=3, blank=True, default=0,
+      verbose_name='Percent matched (%)') # percent of promise matched by employer
+  match_company = models.CharField(max_length=255, blank=True, verbose_name='Employer name' )#  employer name
+  match_received = models.PositiveIntegerField(blank=True, default=0,
+      verbose_name='Amount matched ($)')# total $ amount of match received
```
Comment (aisapatino) [inline sjfnw/fund/models.py]: I might say "Percent match expected" and "Match amount received" or something to clarify that one is a guess and one is the actual amount received.

Also, for `match_expected`, instead of `max_length` we can be more explicit/precise and use `MaxValueValidator`. See the likelihood field (also a percentage) for an example:

[`validators=[MaxValueValidator(100)]`](https://github.com/aisapatino/sjfnw/blob/master/sjfnw/fund/models.py#L273)

(Django docs [here](https://docs.djangoproject.com/en/1.5/ref/validators/))

File: sjfnw/fund/models.py [hunk 2]
```diff
@@ -288,11 +288,16 @@ class Donor(models.Model):
...
 
   # contact info only required if promise is entered
   phone = models.CharField(max_length=15, blank=True)
   email = models.EmailField(max_length=100, blank=True)
-  notes = models.TextField(blank=True, )
+  notes = models.CharField(max_length=100, blank=True)
```
Comment (aisapatino) [inline sjfnw/fund/models.py]: This would correspond to a a change in the database column type, also (currently it's a `longtext`, `CharField` corresponds to a `varchar`). I think it will work without updating the column, but if you're just wanting a smaller widget, it would be safer to change the widget instead. I think 100 is a little stringent as a hard limit for the notes field. For instance if someone wanted to put this as a note: `They are interested in doing an Environmental Justice Giving Project next year. They're not sure about company match so I left it blank.` That's 136 chars.

File: sjfnw/fund/views.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+from __future__ import division
```
Comment (aisapatino) [inline sjfnw/fund/views.py]: is this used?

File: sjfnw/fund/views.py [hunk 4]
```diff
@@ -153,6 +155,8 @@ def _compile_membership_progress(donors):
       progress['promised'] += donor.promised
       donor_data[donor.pk]['next_date'] = datetime.date(2700, 1, 1)
       donor_data[donor.pk]['summary'] += ' Promised $%s.' % intcomma(donor.promised)
+      donor_data[donor.pk]['summary'] += ' Matched $%s.' % intcomma(donor.match_received)
```
Comment (aisapatino) [inline sjfnw/fund/views.py]: `match_received` could go with gift received (`donor.received()` above), with promise maybe `match_expected` would make sense?

File: sjfnw/fund/views.py [hunk 5]
```diff
@@ -153,6 +155,8 @@ def _compile_membership_progress(donors):
...
+      donor_data[donor.pk]['summary'] += ' Total $%s.' % intcomma(donor.match_received + donor.promised)
```
Comment (aisapatino) [inline sjfnw/fund/views.py]: Total expected, perhaps? (To make sure it's clear expected vs. actually received by SJF)

File: sjfnw/fund/views.py [hunk 6]
```diff
@@ -977,6 +981,11 @@ def complete_step(request, donor_id, step_id):
       asked = form.cleaned_data['asked']
       response = form.cleaned_data['response']
       promised = form.cleaned_data['promised_amount']
+      match_expected = form.cleaned_data['match_expected']
+      match_company = form.cleaned_data['match_company']
+
+      match_ratio = match_expected/100
+      match_received = promised * match_ratio
```
Comment (aisapatino) [inline sjfnw/fund/views.py]: We shouldn't be setting `match_received`, right? That's entered by staff when they actually get the match from the company. Also line 987 is what's breaking the tests because `match_expected` could be `None` if they didn't enter any.  You could update it to something like
```
if match_expected:
  match_expected = promised * match_expected // 100 # floor division
```
That will store `match_expected` as an amount instead of a percentage. And you could move it down to inside the block that handles new promises (I'll put a comment there)

File: sjfnw/fund/views.py [hunk 7]
```diff
@@ -1003,6 +1012,9 @@ def complete_step(request, donor_id, step_id):
             donor.phone = phone
           if email:
             donor.email = email
+          donor.match_expected = match_expected
+          donor.match_company = match_company
+          donor.match_received = match_received
```
Comment (aisapatino) [inline sjfnw/fund/views.py]: Don't need to update this here. A gp member should not be able to influence the `match_received` field, that's staff-entered only. Similar to the other `received_` fields

**Comment by aisapatino** [comment]
This looks good! I put in some comments including what I think is breaking most of the tests. You may have done this already - if you click on a 'details' link from the failed checks, you can scroll down and see the test results including the reason for test failures. It's a lot of `TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'` - I commented on the line that's triggering that.

But yeah overall looks really good. I haven't had a chance to try running it locally or to take a close look at the tests, but I will do that later tonight probably.

File: sjfnw/fund/models.py (hunk 1 above)
Comment (tarhata) [inline sjfnw/fund/models.py]: Does `min_value` and `max_value` refer to restricting the percent matched? I was debating whether to impose a max value, but thought there might be a case where an employer might match more than 100%. This might be the case if the employer has charity partnerships with other companies who agree to match an employer match (if that makes sense)

File: sjfnw/fund/models.py (hunk 2 above)
Comment (tarhata) [inline sjfnw/fund/models.py]: Ack! Not sure why I did that. I keep forgetting to change sizes in forms.py . Sorry about thaat.

File: sjfnw/fund/models.py (hunk 1 above)
Comment (aisapatino) [inline sjfnw/fund/models.py]: Oh, good point. Maybe just min value of 0 (so it matches the model
requirements)

In sjfnw/fund/models.py
<https://github.com/aisapatino/sjfnw/pull/738#discussion_r26590388>:

> @@ -288,11 +288,16 @@ class Donor(models.Model):
>    received_afternext = models.PositiveIntegerField(default=0,
>        verbose_name='Received - year after next')
>    gift_notified = models.BooleanField(default=False)
> +  match_expected = models.PositiveIntegerField(max_length=3, blank=True, default=0,
> +      verbose_name='Percent matched (%)') # percent of promise matched by employer
> +  match_company = models.CharField(max_length=255, blank=True, verbose_name='Employer name' )#  employer name
> +  match_received = models.PositiveIntegerField(blank=True, default=0,
> +      verbose_name='Amount matched ($)')# total $ amount of match received

Does min_value and max_value refer to restricting the percent matched? I
was debating whether to impose a max value, but thought there might be a
case where an employer might match more than 100%. This might be the case
if the employer has charity partnerships with other companies who agree to
match an employer match (if that makes sense)

—
Reply to this email directly or view it on GitHub
<https://github.com/aisapatino/sjfnw/pull/738/files#r26590388>.

File: sjfnw/fund/views.py (hunk 3 above)
Comment (tarhata) [inline sjfnw/fund/views.py]: Thanks! I had looked how to deal with integer division to float on stack overflow and the 'from_future_import division` was one suggestion. Didn't realize that that's what `//` meant.

**Comment by tarhata** [comment]
I think I misunderstood what `match_received` is. Since we were talking about the what kind of data to ask from the member (% vs. $ amount), in my head I thought that `match_received` referred to the $ amount that results from % matched x $ Promise.

If that's not the case, then `match_received` should always = 0 until staff manually inputs the amount when the match is actually received, right? Then when staff does confirm that, then the `match_received` amount should be added to `donor.received()` and ( `progress['received']`.

I'm going to ask for more clarification regarding the terminology used for the entire giving project process and what kind of data they'd like to see on the admin site.

**Comment by aisapatino** [comment]
Ah ok. I was thinking of it as the received counterpart to the expected field, to track when the money has been received by SJF. We can calculate between percentage and amount as needed so I don't think we need to store both. So yep, it would be 0 until staff enters something. By updating the `donor.received()` method, I think `progress['received']` should be updated automatically because it uses that method to get the amount. But we should write at least one test to confirm that.

**Comment by tarhata** [comment]
Decided to just commit what I have so far before getting dinner. I changed  match_expected to be an amount since it didn't seem to make sense to make it a percentage initially if it would ultimately be stored as an amount. This is especially true if we want members to have the ability to change match amounts if they forget or accidentally enter the wrong number.

I'm still trying to figure out how to change the progress display and how to dynamically change the promise amount when the match amount changes. For example, if a member reduces the match amount, then the promise should always update.

I need help with writing tests, so if you have time to help, please do!

**Comment by aisapatino** [comment]
I think theres an easier way to do it - the actual `promised` field should not change, but we can add a method to the `Donor` model to compute the total promise including match. I'll write it up when I get home (between 8:30 and 9 probably). I can write some tests later tonight too :)

**Comment by tarhata** [comment]
O yeah!  I forgot that I could write a method for the `Donor` model that acts similarly to `donor.received()`. I'll try to write it and commit it for your reviews. And those tests... :cry: I just don't get it yet!!

**Comment by aisapatino** [comment]
Okay, I'm going to check this out now and see if I can get the tests passing. If you're still working on anything on this branch, let me know, and we can coordinate pushing/merging/rebasing.

**Comment by tarhata** [comment]
I'm still working on it! I fixed the `total_promised` to return `self.promised` if `match_expected` is null. There were also some errors from bad code I left in, so I removed some code in `models.py`

**Comment by aisapatino** [comment]
Okay cool. I'm just working in tests right now so there shouldn't be to much trouble putting them together. Just push those fixes up whenever you're at a break point.

**Comment by aisapatino** [comment]
I just pushed my changes; there's a test for each scenario (existing tests cover "no match entered", then there's one for only company, only expected amount and both)

Do a `git pull --rebase` to update your local branch. (`--rebase` avoids the merge commit)

File: sjfnw/fund/forms.py [hunk 8]
```diff
@@ -124,6 +124,10 @@ class StepDoneForm(forms.Form):
       required=False, min_value=0,
       error_messages={'min_value': 'Promise amounts cannot be negative'},
       widget=forms.TextInput(attrs={'size':10}))
+  match_expected = forms.IntegerField(required=False, label='Amount matched',
+                                      min_value=0, widget=forms.TextInput(attrs={'size':'1'}))
```
Comment (aisapatino) [inline sjfnw/fund/forms.py]: I'd make this a little bigger - for me it only fits 2 digits. Should probably fit 4, maybe 5 just to be safe. Looks like size of 4 works.

File: sjfnw/templates/fund/includes/donor_details.html [hunk 9]
```diff
@@ -63,13 +63,12 @@
       {% elif d.donor.promised != None %}
         <p>
           {% if d.donor.promised > 0 %}
-            Promised: ${{ d.donor.promised|intcomma }}
+            Promised: ${{ d.donor.total_promised|intcomma }}
```
Comment (aisapatino) [inline sjfnw/templates/fund/includes/donor_details.html]: I would lean toward leaving the original value here because it seems like it might be confusing if the actual amount promised by the contact isn't shown anywhere. What about adding a line below it saying 'Expected employer match' or something (inside a `{% if donor.match_expected %}` check)

So instead of:
```
Asked. Promised $2,678 (includes expected match of $1,078.)
...
Promised: $2,678
```
It could say
```
Asked. Promised $2,678 (includes expected match of $1,078.)
...
Promised: $1,600
Expected employer match: $1,078
```

File: sjfnw/fund/forms.py (hunk 8 above)
Comment (tarhata) [inline sjfnw/fund/forms.py]: Really?! On my browsers (Chromium and Chrome) it fits 5 digits! I'll just make the input width match the length of the `Amount matched` label.

File: sjfnw/fund/admin.py [hunk 10]
```diff
@@ -38,7 +38,7 @@ def ship_progress(obj):
           str(p['received_total']) + '</td><td style="width:25%;padding:1px">' +
           str(p['received_this']) + ', ' + str(p['received_next']) +
           ', ' + str(p['received_afternext']) + '</td></tr></table>')
-ship_progress.short_description = 'Estimated, promised, received, rec. by year'
+ship_progress.short_description = 'Estimated, total promised, received, rec. by year'
```
Comment (aisapatino) [inline sjfnw/fund/admin.py]: I think we might want to leave this as just the amount promised by the donor. The columns are set up like this
`Promised | Received - this | Received - next | Received - year after next | Match expected | Match received`

So having `match_expected` included in promised seems redundant since it's shown on its own too. Also potentially confusing because the 3 'received' columns after 'promised' all apply to the donor only.  What do you think?

File: sjfnw/fund/admin.py [hunk 11]
```diff
@@ -149,8 +149,9 @@ class DonorInline(admin.TabularInline): #membership
   max_num = 0
   can_delete = False
   readonly_fields = ('firstname', 'lastname', 'amount', 'talked', 'asked',
-                     'promised')
-  fields = ('firstname', 'lastname', 'amount', 'talked', 'asked', 'promised')
+                     'promised', 'total_promised')
+  fields = ('firstname', 'lastname', 'amount', 'talked', 'asked', 'promised',
+            'match_expected', 'match_received')
```
Comment (aisapatino) [inline sjfnw/fund/admin.py]: I don't think we need the match fields on the inline, unless they specified. We don't even have gift received fields on the inline currently. (The inline only shows when viewing a single membership)

File: sjfnw/fund/admin.py [hunk 12]
```diff
@@ -247,8 +248,9 @@ def approve(self, request, queryset): #Membership action
...
 
 class DonorA(admin.ModelAdmin):
-  list_display = ('firstname', 'lastname', 'membership', 'amount', 'talked', 'asked',
-                  'promised', 'received_this', 'received_next', 'received_afternext')
+  list_display = ('firstname', 'lastname', 'membership', 'amount', 'talked',
+                  'asked', 'total_promised', 'received_this', 'received_next',
+                  'received_afternext','match_expected', 'match_received')
   list_filter = ('membership__giving_project', 'asked', PromisedBooleanFilter,
                  ReceivedBooleanFilter)
   list_editable = ('received_this', 'received_next', 'received_afternext')
```
Comment (aisapatino) [inline sjfnw/fund/admin.py]: Should add `match_expected` and `match_received` to `list_editable`

File: sjfnw/templates/fund/includes/donor_details.html (hunk 9 above)
Comment (tarhata) [inline sjfnw/templates/fund/includes/donor_details.html]: Agreed. I thought this was confusing too. Since I changed the admin site to list 'Total Promise', maybe it should be:
```
Asked. Total Promised $2,678 (includes expected match of $1,078.)
...
Promised: $2,678
Expected employer match: $1078
```
However, should members be able to change `Promised` just like  `Match expected ($)`  in edit contact? Or should I remove the ability to do either and force members to contact staff for changes?

File: sjfnw/fund/forms.py (hunk 8 above)
Comment (aisapatino) [inline sjfnw/fund/forms.py]: Weird! I'm on Chrome too. I'm not very familiar with `size` but it looks like it's meant to be [the number of characters that fit](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Input#attr-size). But it definitely fits more than one character, so not sure what the deal is. I guess better too big than too small.

File: sjfnw/templates/fund/includes/donor_details.html (hunk 9 above)
Comment (aisapatino) [inline sjfnw/templates/fund/includes/donor_details.html]: I like the summary like that. In the details view, since space is less of a concern, I think it would be nice for them to be able to see the amount that the person promised to give, too - not just the total including the match.  I can imagine it'd be confusing if a contact promised $1,600 and then that number doesn't appear anywhere in the donor details. Could be verbose and do
```
Asked. Total promised: $2,678 (includes expected match of $1,078)
...
Total promised: $2,678
Promised by contact: $1,600
Expected employer match: $1,078
```
Or something along those lines, just to be explicit. But who knows, I've stared at it so long I can't tell what makes sense :grimacing: 

They should not be able to change the promise amount at this point - there's some complicated mess involving the way steps are set up. Editing match amount is okay, it shouldn't affect anything else, so either way is fine.

**Comment by tarhata** [comment]
Was playing around with this latest branch some more last night and found a **huge bug!!** If a contact's response is marked `Unsure` and saved, if you try to complete the step afterwards in `complete_step.html` the entire page freezes. The error shown in the terminal is:

```
error: http://localhost:8080/static/js/fund_personal.js (148): Uncaught TypeError: Cannot read property 'match' of undefined
```

Looked at inspect element in Chrome and it corresponds to this line in `fund_personal.js`:
`var donorId = response.id.match(/\d+/);`

**Comment by aisapatino** [comment]
I probably introduced that bug when I renamed some element ids. Good find! I'll take a look.

**Comment by aisapatino** [comment]
Fixed. The javascript is a bit of a mess overall; in some cases I was passing in an html element and in others a jquery selection.  jquery selection does not have an `id` property, which is why it was breaking on that line.

**Comment by tarhata** [comment]
Damn that was fast! I'll look through the app more after my dentist appt,
but let me know if there's anything I missed.

**Comment by aisapatino** [comment]
Dentist day for me too! I have a 12:30 appt and then a 3pm vet appt but
I'll take another look in between.

**Comment by aisapatino** [comment]
@tarhata it looks like you've done everything - i checked off the remaining items in #703. does that look right? anything you know of that still needs to be done?

i'm going to test out all the home page forms to make sure i didn't break anything else. then just give me the okay and i can deploy it :+1:

**Comment by tarhata** [comment]
Awesome! I didn't find anything, so I think we're good to go. I had a mini
usability test with my partner too (she's the one who found the bug, hehe),
so we were pretty thorough 😊

**Comment by aisapatino** [comment]
Sounds good! Merging
